
<table style="border-collapse:collapse;" class=table_9950 border=1>
<col width="360">
<col width="200">
<thead>
<tr style="background-color:#f2f2f2;">
  <th id="tableHTML_header_1">PARAMETER</th>
  <th id="tableHTML_header_2">VALUE</th>
</tr>
</thead>
<tbody>
<tr style="background-color:#e6f0ff;">
  <td id="tableHTML_column_1">Genotype filename</td>
  <td id="tableHTML_column_2">example-genotype.tbl</td>
</tr>
<tr style="background-color:#f2f2f2;">
  <td id="tableHTML_column_1">Phenotype filename</td>
  <td id="tableHTML_column_2">example-phenotype.tbl</td>
</tr>
<tr style="background-color:#e6f0ff;">
  <td id="tableHTML_column_1">Significance level (Genome-wide significance level)</td>
  <td id="tableHTML_column_2">0.05</td>
</tr>
<tr style="background-color:#f2f2f2;">
  <td id="tableHTML_column_1">Correction method (Bonferroni or FDR)</td>
  <td id="tableHTML_column_2">Bonferroni</td>
</tr>
<tr style="background-color:#e6f0ff;">
  <td id="tableHTML_column_1">GWAS model (Full or Naive)</td>
  <td id="tableHTML_column_2">Naive</td>
</tr>
<tr style="background-color:#f2f2f2;">
  <td id="tableHTML_column_1">NBest (Number of best-ranked SNPs to be reported)</td>
  <td id="tableHTML_column_2">4</td>
</tr>
<tr style="background-color:#e6f0ff;">
  <td id="tableHTML_column_1">Filtering (TRUE or FALSE)</td>
  <td id="tableHTML_column_2">TRUE</td>
</tr>
<tr style="background-color:#f2f2f2;">
  <td id="tableHTML_column_1">MIND Filter (Individual with missing genotype)</td>
  <td id="tableHTML_column_2">0.1</td>
</tr>
<tr style="background-color:#e6f0ff;">
  <td id="tableHTML_column_1">GENO Filter (SNPs with missing genotype)</td>
  <td id="tableHTML_column_2">0.1</td>
</tr>
<tr style="background-color:#f2f2f2;">
  <td id="tableHTML_column_1">MAF Filter (Minor allele frequency)</td>
  <td id="tableHTML_column_2">0.01</td>
</tr>
<tr style="background-color:#e6f0ff;">
  <td id="tableHTML_column_1">HWE Filter (Hardy-Weinberg test)</td>
  <td id="tableHTML_column_2">1e-10</td>
</tr>
</tbody>
</table>